django/core/management/commands/compilemessages.py

```diff
@@ -65,6 +71,7 @@ def handle(self, **options):
 
         # Walk entire tree, looking for locale directories
         for dirpath, dirnames, filenames in os.walk('.', topdown=True):
+            dirnames[:] = [d for d in dirnames if d not in ignore]
             for dirname in dirnames:
                 if dirname == 'locale':
```

Comment:
You can avoid the double iteration over `dirnames` and list creation by simply adjusting the if below.

```suggestion
                 if dirname == 'locale' and dirname not in ignore:
```

This change made me realize you probably want to consider `dirpath` unless I'm missing something as only `dirname == 'locale'` will be considered anyway? At least that's what the reporter of the ticket seems to be trying to do by passing `--ignore .tox`. You might want to have a look at [how `makemessages --ignore` is implemented](https://github.com/django/django/blob/4bc008028c33cea5307a08c401778fe503e063a4/django/core/management/commands/makemessages.py#L457-L477).

Reply:
You are right, we can definitely avoid the double iteration. I think we still want to prune the `dirnames` list on a matching `dirname` though otherwise it will recurse into that dir to look for a nested locale directory.
```
./
 |- locale/ 
 |- foo/
  |- locale/
```

We can get the same result checking the path like you suggested, but it would add extra iterations of walk as it would still navigate further down the directory.